### ISSUE TITLE
Add the ability to make multiline suggestions/notes

### ISSUE THREAD

**Issue by RDambrosio016** [brendanzab/codespan#293]
Hello!

I have been working on a large project and a lot of the errors have suggestions attached to them, however, i would greatly appreciate being able to make suggestions like rustc_errors can do. Example:
```
error: unknown start of token: \u{37e}
 --> main.rs:2:30
  |
2 |    println!("Hello, world!");
  |                             ^
help: Unicode character ';' (Greek Question Mark) looks like ';' (Semicolon), but it is not
  |
2 |    println!("Hello, world!");
  |                             ^
```
Being able to make suggestions which point to some lines in the source code but change it and label the changes would be a great feature, however, thinking of a good API to provide this will not be an easy task. Another thing i have considered is potentially exposing the renderer API to allow users to tack on their own render logic, which would help me immensely.

**Comment by brendanzab** [comment]
Yeah, this has been requested before (See #285)! Perhaps this is something we should have a look at.

For this specific instance, I'm wondering if using the notes section would help. Eg.

```
error: unknown start of token: \u{37e}
  ┌─ main.rs:2:30
  |
2 |    println!("Hello, world!");
  |                             ^
  = Unicode character ';' (Greek Question Mark) looks like ';' (Semicolon), but it is not
```

But I realise there are other instances where this is not as helpful, say with lifetime diagnostics, where you have to step through a sequence of related locations.

It would also be good to think about how this maps onto the language server protocol, but we might not want to limit ourselves by it. Eg. I tend to find lifetime errors easier to understand in the terminal, rather than using rust-analyzer.

**Comment by RDambrosio016** [comment]
As with most pretty error stuff, the logic will be lost in the LSP, which is not necessarily bad, i agree with you in saying cli is way better for some errors, i dont think LSP should limit codespan when it comes to features. The example was the first i could think of, it is obviously very simplistic.

**Comment by Johann150** [comment]
I have not looked at the commits that were made, but maybe #176 could be interesting in this context?

**Comment by RDambrosio016** [comment]
Allowing users to write their own render logic and plug into codespan's renderer would be infinitely helpful when it comes to more specific diagnostics

**Comment by Aapuji** [comment]
Hello, I was wondering what the status is for this issue? Thanks

**Comment by kaleidawave** [comment]
I think some of this has been implemented and discussed in the spade fork: https://github.com/brendanzab/codespan/issues/372 other that there is no open PR implementing this.

I may open a fresh issue summarising why this is needed over notes and whether an LSP implementation can reuse this.

**Comment by kaleidawave** [comment]
The description of this issue could be improved to clarify whether it means
1. The `notes` field to allow `\n` characters (I think this already works? not sure)
2. Allowing [`Label`](https://docs.rs/codespan-reporting/latest/codespan_reporting/diagnostic/struct.Label.html) to include multiple ranges. I believe I #346 implements this
3. Adding a suggestions system see #372
